tfhe/src/shortint/engine/server_side/div_mod.rs

```diff
@@ -39,17 +39,27 @@ impl ShortintEngine {
         Ok(())
     }
 
-    // by convention smart operations take mut refs to their inputs, even if they do not modify them
-    #[allow(clippy::needless_pass_by_ref_mut)]
     pub(crate) fn smart_div(
         &mut self,
         server_key: &ServerKey,
         ct_left: &mut Ciphertext,
         ct_right: &mut Ciphertext,
     ) -> EngineResult<Ciphertext> {
-        let mut result = ct_left.clone();
-        self.smart_div_assign(server_key, &mut result, ct_right)?;
-        Ok(result)
+        if !server_key.is_functional_bivariate_pbs_possible(ct_left, ct_right) {
+            if ct_left.message_modulus.0 + ct_right.degree.0 <= server_key.max_degree.0 {
+                self.message_extract_assign(server_key, ct_left)?;
+            } else if ct_right.message_modulus.0 + (ct_left.degree.0 + 1) <= server_key.max_degree.0
+            {
+                self.message_extract_assign(server_key, ct_right)?;
+            } else {
+                self.message_extract_assign(server_key, ct_left)?;
+                self.message_extract_assign(server_key, ct_right)?;
+            }
+        }
```

Comment:
Why did you do something more complex here than for bitand/bitor which are also bivariate PBSes

Also the correlations of the ifs to the message extract are weird:

- `if ct_left.message_modulus.0 + ct_right.degree.0 <= server_key.max_degree.0`
 -> you message extract ct_left, but you checked ct_right degree not ct_left
    and you used ct_left message modulus, but after message_extract modulus, ct_right degree would not have changed, and ct_left degreee would be message_modulus - 1
to me, the potential issue is not fixed 

same idea for the `else if`

Reply:
I just copied what is done in `smart_div_assign` to `smart_div`  so that inputs are cleaned instead of their clones

The behavior (except the potential cleaning of the inputs) should be the same as before
I suggest we fix this in another PR

WDYT?

Reply:
Ok if it comes from smart_div_assign then we can merge and think more about this strange thing

Reply:
https://github.com/zama-ai/tfhe-rs-internal/issues/250